src/backend/distributed/executor/multi_router_executor.c

```diff
@@ -591,3 +646,288 @@ RouterExecutorEnd(QueryDesc *queryDesc)
 	queryDesc->estate = NULL;
 	queryDesc->totaltime = NULL;
 }
+
+/*
+ * Walks each TargetEntry of the query, evaluates sub-expressions without Vars.
+ */
+void ExecuteFunctions(Query *query)
+{
+	CmdType commandType = query->commandType;
+	ListCell *targetEntryCell = NULL;
+	Node *modifiedNode = NULL;
+
+	/* DELETE has no targetList */
+	if (!(commandType == CMD_INSERT) && !(commandType == CMD_UPDATE))
+	{
+		return;
+	}
+
+	foreach(targetEntryCell, query->targetList)
+	{
+		TargetEntry *targetEntry = (TargetEntry *) lfirst(targetEntryCell);
+
+		/* performance optimization for the most common cases */
+		if (IsA(targetEntry->expr, Const) || IsA(targetEntry->expr, Var))
+		{
+			continue;
+		}
+
+		modifiedNode = PartiallyEvaluateExpression((Node *) targetEntry->expr);
+
+		targetEntry->expr = (Expr *) modifiedNode;
+	}
+}
+
+/*
+ * Walks the expression, evaluates sub-expressions which don't contain Vars.
+ *
+ * The flow of control is a little convoluted to prevent ourselves from needing to walk
+ * parts of the tree multiple times. When we walk a node which contains multiple
+ * sub-trees we recurse and remember which parts of the tree contained a Var. If some
+ * did and some didn't we evaluate the parts of the tree which didn't. If none of the
+ * sub-trees contained a Var we do nothing besides reporting that fact.
+ *
+ * At the top-level of the expression we evaluate everything if no Vars were found.
+ */
+static Node * PartiallyEvaluateExpression(Node *expression)
+{
+	ExpressionWalkerState unused;
+	InitializeWalkerState(&unused);
+	unused.isTopLevel = true;
+
+	return PartiallyEvaluateExpressionWalker(expression, &unused);
+}
+
+
+static Node * PartiallyEvaluateExpressionWalker(Node *expression,
```

Comment:
I don't think all this code should live in the router executor. How about utils/citus_expr.c or something?